ArgonKey board: Adjust DTS support and add sample test

avisconti opened the pull request:
Extending the ArgonKey support with:

   - sample code to test all on-board sensors
   - split LSM6DSL dts binding for I2C and SPI.
   - Add in board dts all I2C sensors and LSM6DSL SPI sensor.
   - STM32 I2C interrupt handling enabled by default

avisconti committed:
samples/boards: Add ArgonKey test

Test ArgonKey board. It provides an example of how to test
all the sensors on-board. It currently enables by default
following sensors:

   LPS22HB pressure/temp
   HTS221 humidity
   LSM6DSL accel/gyro plus LIS2MDL magn connected to it.
   VL53L0x proximity

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
boards: stm32: argonkey: Add dts support to sensors

This patch adds dts support to lps22hb/hts221/vl53l0x
I2C sensors as well as lsm6dsl SPI sensor.

Since some info, like gpio for irq triggering or SPI bus
characteristics, may be provided through dts, they need
to be made optional in the LSM6DSL driver by usage of
macros like HAS_DTS_SPI_PINS or HAS_DTS_SPI_DEVICE.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
Area is board 
@erwango

erwango commented:
@avisconti : Don't you miss CONFIG flags to be configured for good sensor configuration?
Besides, you might want to add 96b_argonkey to sample/sensor/vl53l0x

avisconti commented:
@erwango 
Concerning the argonkey I was going to add a test here: samples/boards/argonkey

erwango commented:
> Concerning the argonkey I was going to add a test here: samples/boards/argonkey

Sure go ahead!

ydamigos commented on boards/arm/96b_argonkey/96b_argonkey_defconfig:
```diff
@@ -23,6 +23,21 @@ CONFIG_I2C=y
+CONFIG_LSM6DSL_GPIO_DEV_NAME="GPIOB"
+CONFIG_LSM6DSL_GPIO_PIN_NUM=1
+CONFIG_LSM6DSL_SENSORHUB=y
+CONFIG_LSM6DSL_EXT0_LIS2MDL=y
```
I am not familiar with 96b_argonkey but shouldn't all the following being moved in the 96b_argonkey's test sample? 
```
CONFIG_SENSOR=y
CONFIG_HTS221=y
CONFIG_LPS22HB=y
CONFIG_LSM6DSL=y
CONFIG_VL53L0X=y
```
And the following in the board's Kconfig.defconfig?
```
CONFIG_LSM6DSL=y
CONFIG_LSM6DSL_SPI=y
CONFIG_LSM6DSL_SPI_MASTER_DEV_NAME="SPI_2"
CONFIG_LSM6DSL_SPI_SELECT_SLAVE=1
CONFIG_LSM6DSL_TRIGGER_GLOBAL_THREAD=y
CONFIG_LSM6DSL_TRIGGER=y
CONFIG_LSM6DSL_GPIO_DEV_NAME="GPIOB"
CONFIG_LSM6DSL_GPIO_PIN_NUM=1
CONFIG_LSM6DSL_SENSORHUB=y
CONFIG_LSM6DSL_EXT0_LIS2MDL=y
```

erwango replied:
@ydamigos  Exact. Configuration should be done at board level (use `Kconfig.defconfig` as much as possible).
Then, enabling the sensors (SENSOR, LSM6DSL, ...) should be done at application level (`sample/.../.../prj.conf` in this case)

avisconti replied:
Thx guys!
I will follow your indications.

erwango commented:
@avisconti : Since this PR has a new scope, please update titel and description.
Also, can you base your patchset on #6966 ? It should ease your board configuration and will help to validate this PR.

avisconti committed:
board: argonkey: add LSM6DSL configuration in Kconfig.defconfig

Use the named choices feature introduced by PR #6966
and configure LSM6DSL on ArgonKey board.
The two named choices introduced in LSM6DSL are:

    - LSM6DSL_BUS_TYPE (default LSM6DSL_I2C)
    - LSM6DSL_TRIGGER_MODE (default LSM6DSL_TRIGGER_NONE)

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@erwango @ydamigos 
I resubmitted the PR following your suggestions.
But I'm not sure how to pass the build CI test. Basically it seems to fail when compiling
this sample for frdm_k64f board. This is obviously true. But it happens for other samples
too, like samples/sensor/hts221.

So, what should I do?
Should I avoid to enable the sensors in the sample?
Other better methods? Thx in advance for putting some light on this topic.

Thx,
Armando

erwango commented on samples/boards/argonkey/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+  name: ArgonKey test
+tests:
+  test:
+    harness: sensor
```
You should add here:
        platform_whitelist: 96b_argonkey

dbkinder commented on samples/boards/argonkey/README.rst:
```diff
@@ -0,0 +1,53 @@
+************
+
+This sample just requires the ArgonKey board. The user may select
+or unselect the sensors from prj.conf.
```
Change to say:

     ``samples/boards/argonkey/prj.conf``

If I unselect one of the sensors, will the build fail or just not output meaningful data for that sample because the sensor is disabled?

avisconti replied:
Everything is under ifdefs, so it will build fine and just avoid to configure/print sensor drivers not present.

ydamigos commented on boards/arm/96b_argonkey/Kconfig.defconfig:
```diff
@@ -37,4 +37,34 @@ config SPI_2
+config LSM6DSL_GPIO_PIN_NUM
+	default 1
+
+config LSM6DSL_SENSORHUB
```
warning: LSM6DSL_SENSORHUB (defined at /media/projects/zephyr/boards/arm/96b_argonkey/Kconfig.defconfig:62) defined without a type

avisconti replied:
Yes, Yannick.
Actually those depends by #7028 which is still under discussion.
I guess everything will be merged smoothly. What do you think?
Should I instead remove these two stuff and do another commit later on?
Or maybe split this commit into two, but keeping them in this PR itself?

ydamigos replied:
I would mark this PR with a DNM label and wait for #7028 before merging this one.

avisconti replied:
Sure!
I don't think I have permission to do it, have I?
@erwango @galak  can you help me?

ydamigos replied:
I added the DNM label.

ydamigos commented on boards/arm/96b_argonkey/Kconfig.defconfig:
```diff
@@ -37,4 +37,34 @@ config SPI_2
+config LSM6DSL_SENSORHUB
+	default y
+
+config LSM6DSL_EXT0_LIS2MDL
```
warning: LSM6DSL_EXT0_LIS2MDL (defined at /media/projects/zephyr/boards/arm/96b_argonkey/Kconfig.defconfig:65) defined without a type

avisconti replied:
see above.

ydamigos reviewed (changes requested):
Where are the options LSM6DSL_SENSORHUB and LSM6DSL_EXT0_LIS2MDL defined?

avisconti commented on boards/arm/96b_argonkey/Kconfig.defconfig:
```diff
@@ -37,4 +37,34 @@ config SPI_2
+config LSM6DSL_SPI_SELECT_SLAVE
+	default 1
+
+choice LSM6DSL_TRIGGER_MODE
```
@ydamigos 
Well, I am not a Kconfig expert either.
I just copied the way explored by @erwango in PR #6966. There, if you see, the choice inside the driver Kconfig has been modified using a 'named choice' (i.e. a choice with symbol). In this way it seems to be possible to change the default in board Kconfig.defconfig.

ydamigos commented on samples/boards/argonkey/src/main.c:
```diff
@@ -0,0 +1,350 @@
+	}
+
+}
+
```
Please remove the empty lines.

avisconti replied:
ACK

avisconti commented:
@dbkinder @ydamigos 
Re-pushed following your indications

ydamigos commented on samples/boards/argonkey/src/main.c:
```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright (c) 2012-2014 Wind River Systems, Inc.
```
Please fix the copyright.

avisconti replied:
ooops.... ACK!

ydamigos commented on samples/boards/argonkey/src/main.c:
```diff
@@ -0,0 +1,343 @@
+	int cnt = 0;
+	char out_str[64];
+
+#ifdef CONFIG_BOARD_96B_ARGONKEY
```
This is a sample for the 96b_argonkey board why do you add `#ifdef CONFIG_BOARD_96B_ARGONKEY` block around the code which configures the board's leds?

avisconti replied:
Because I built previously the same code for other platform too (e.g. nucleo f411). I forgot to delete it now that I decided to make it available for testing ArgonKey.
ACK

ydamigos reviewed (changes requested):
Almost there! Some minor comments.

dbkinder commented on samples/boards/argonkey/README.rst:
```diff
@@ -0,0 +1,74 @@
+ #. *LIS2MDL* magnetometer (attached to *LSM6DSL*)
+
+The last line displays a counter of how many trigger interrupts
+has been receveid.  It is possible to display the sensor data
```
change to ``received``

dbkinder commented on samples/boards/argonkey/README.rst:
```diff
@@ -0,0 +1,74 @@
+
+The last line displays a counter of how many trigger interrupts
+has been receveid.  It is possible to display the sensor data
+read for each trigger just enabling the **ARGONKEY_TEST_LOG** macro.
```
change ``just`` to ``by``

dbkinder commented on samples/boards/argonkey/README.rst:
```diff
@@ -0,0 +1,74 @@
+References
+**********
+
+ - ArgonKey board.
```
List item should not be indented (remove leading space before the ``-``).   This should probably be a link to our ArgonKey board documentation, so change to:

    - :ref:`96b_argonkey`

avisconti replied:
OK.
Do I have to change also the list:
 #. *LPS22HB* baro/temp
 #. *HTS221* humidity
 #. *LSM6DSL* accel
 #. *LSM6DSL* gyro
 #. *LIS2MDL* magnetometer (attached to *LSM6DSL*)

Correct?

avisconti replied:
Well, I just did it!

avisconti commented on samples/boards/argonkey/README.rst:
```diff
@@ -0,0 +1,74 @@
+
+ .. code-block:: c
+
+  #ifdef CONFIG_HTS221
```
ok

avisconti commented on samples/boards/argonkey/README.rst:
```diff
@@ -0,0 +1,74 @@
+    <repeats endlessly every 2s>
+
+In this example the output is generated polling the sensor every
+2s. Sensors data is printed in following order:
```
ack

dbkinder reviewed (approved):
LGTM, thanks.

avisconti committed:
boards: stm32: argonkey: Enable STM32 I2C interrupt support

Enable STM32 I2C interrupt support for faster response

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
Guys, thanks for reviewing.
In the meantime I also found solution to some issues that was preventing enabling few sensors.
So, I repushed the everything with few minor changes:
   - I enabled STM32 I2C interrupt handling (separate commit)
   - In the test I'm enabling also LPS22HB and VL53L0X

Pls note that this PR is now depending by the following two PR and should be merged only after them:
PR #7374 #7401

avisconti commented:
@erwango 
Do you agree with the changes?

erwango commented on samples/boards/argonkey/src/main.c:
```diff
@@ -0,0 +1,340 @@
+
+/* #define ARGONKEY_TEST_LOG 1 */
+
+#define LED0_GPIO_NAME		LED0_GPIO_CONTROLLER
```
Can't we use directly *_GPIO CONTROLLER and *_GPIO_PIN?

avisconti replied:
I think so. I will change it right away.

erwango commented on boards/arm/96b_argonkey/Kconfig.defconfig:
```diff
@@ -40,4 +40,34 @@ config SPI_2
+	default LSM6DSL_SPI
+endchoice
+
+config LSM6DSL_SPI_MASTER_DEV_NAME
```
Isn't it provided from DTS?

avisconti replied:
Is it working fine the SPI from dts? 
I think I already asked this to you, but maybe I found some issue.
I can retry as I do not remember. Will I need to modify the dts.fixup file I guess.

erwango replied:
It's working for some parts such as BT controller for instance.
Yes, please let us know the status.

ydamigos replied:
Yes, it works. I have also used DT for SDHC over SPI.

erwango replied:
This might be because you used "st,lsm6dsl" as compatible, which will call i2c-device.yaml.
You might need to define a new compatible "st,lsm6dsl-spi" (or similar. i don't know how this kind of issue is handled in Linux) with matching "st,lsm6dsl-spi.yaml" that includes spi-device.yaml.
If this work, let's rename "st,lsm6dsl" to "st,lsm6dsl-i2c" (if acceptable from Linux point of view)

avisconti replied:
Thanks @erwango 
It seems exactly the reason why I'm getting the error. Sorry, I didn't know exactly how this dts binding was working in Zephyr. So, the plan now is to provide the new st,lsm6dsl-spi bindings and modify the lsm6dsl to get configufartions from dts. Concerning the fix to lsm6dsl-i2c I would skip this for the moment as it can broke other parts of the code.

avisconti replied:
@erwango @ydamigos 
I'm about to push a new PR which provides SPI configuration for LSM6DSL in dts.
In particular following info are provided through dts:

   CONFIG_LSM6DSL_SPI_SELECT_SLAVE
   CONFIG_LSM6DSL_SPI_MASTER_DEV_NAME
   CONFIG_LSM6DSL_SPI_BUS_FREQ   
   CONFIG_LSM6DSL_GPIO_DEV_NAME
   CONFIG_LSM6DSL_GPIO_PIN_NUM

Please note that the latter two are related to triggers (and not to SPI), but the other board that is using LSM6DSL is disco_l475_iot1 and it is NOT using triggers for LSM6DSL, in fact it builds fine.
Pls provide to me your comments about it.

erwango reviewed (changes requested):
Few comments.
One question: can't we get LSM6DSL data from dts?

avisconti committed:
dts: bindings: Add SPI yaml file for LSM6DSL sensor

The LSM6DSL accel/gyro sensor can be accessed through
SPI bus. So the required configuration can be passed through
dts file.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@erwango @ydamigos 
At the end I also changed the LSM6DSL binding for I2C using "st,lsm6dsl-i2c"
This is done in last commit. Can you possibly cross verify it on discovery?
Thanks!

avisconti commented:
Mmmmmh!
I noticed now following things available in Zephyr:
---
There are two new Kconfig symbols, CONFIG_HAS_DTS_GPIO and
CONFIG_HAS_DTS_GPIO_DEVICE, which respectively allow declaring that the
GPIO driver supports device tree, and that drivers which need GPIOs do as
well. Currently, only the STM32 and MCUX GPIO drivers select
CONFIG_HAS_DTS_GPIO.
---

In this PR I instead use HAS_DTS_SPI_DEVICE.
This is a mistake. I will change it accordingly... Sorry!

avisconti commented:
@erwango 
I think I figured it out. I am now checking in LSM6DSL Kconfig the HAS_DTS_SPI_PINS macro.
To be honest I see some potential issues here which we may need to address in the future, like the fact that we may need to actually separate between cs-gpios, reset-gpios,  irq-gpios.
For the moment I'm using the disco_l475_iotl strategy.

erwango reviewed (changes requested):
Looks good! 
One last comment, can you use actual board (96b_argonkey) name for sample directory?

avisconti commented:
ACK

avisconti commented:
Done!

avisconti commented:
Would be nice to have this PR merged in v1.12. Is it closing today the merging window?

MaureenHelm commented on samples/boards/96b_argonkey/README.rst:
```diff
@@ -0,0 +1,75 @@
+
+ .. code-block:: bash
+
+    $ cd samples/boards/argonkey
```
Use zephyr-app-commands instead. See samples/hello_world/README.rst for an example

avisconti replied:
not clear to me. Why not using bash as a code block here. These are really bash shell commands

avisconti replied:
ok, almost done.

MaureenHelm commented on samples/boards/96b_argonkey/src/main.c:
```diff
@@ -0,0 +1,335 @@
+	printk("ArgonKey test!!\n");
+
+#ifdef CONFIG_LPS22HB
+	struct device *baro_dev = device_get_binding("LPS22HB");
```
Use the device name from dts rather than hardcoding the string. Do this for the other sensors too.

MaureenHelm commented:
> Would be nice to have this PR merged in v1.12. Is it closing today the merging window?

Yes, the merge window will close today

avisconti commented on boards/arm/96b_argonkey/dts.fixup:
```diff
@@ -0,0 +1,22 @@
+/* This file is a temporary workaround for mapping of the generated information
```
ack

avisconti commented:
@MaureenHelm 
It's a pitty to not being able to make it for v1.12. Unfortunately I really need to leave now for personal reason. But it seems that I need a couple of hours to fix what you have highlighted. I'll do that on Monday. Have a nice weekend

avisconti commented:
Actually I may come back in 1h and finish the work.

avisconti commented on samples/boards/96b_argonkey/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+
```
ack, done

avisconti commented on samples/boards/96b_argonkey/README.rst:
```diff
@@ -0,0 +1,75 @@
+
+This sample just requires the ArgonKey board.
+The user may select or unselect the sensors from
+:file:`samples/boards/argonkey/prj.conf`.
```
ack, done
